docs/howtos/testing-a-rust-component.md

```diff
@@ -0,0 +1,146 @@
+  code are not in Rust, that's sometimes not possible. The best idiomatic Rust API for a feature
+  is not necessarily the best API for consuming it over an FFI boundary!
+
+* Rust's builtin assertion macros are pretty spartan; we use the [more_asserts](https://crates.io/crates/more_asserts)
+  for some additional helpers.
+
+* Rust's strict typing can make test mocks difficult. If there's something you need to mock out in tests,
+  make it a Trait and use the [mockiato](https://crates.io/crates/mockiato) crate to mock it out.
+
+The Rust tests for a component should be runnable via `cargo test`.
+
+### FFI Layer code
+
+We currently do not test the FFI-layer Rust code for our components, since it's generally a very thin
+wrapper around the underlying (and in theory well-tested!) Rust component code. If you find yourself
+adding a particularly complex bit of code in an FFI-layer crate, add unittests in the same style as
+for other Rust code.
+
+(Editor's note: I remain hopeful that one day we'll autogenerate most of the FFI-layer code, and in
+such a world we don't need to invest in tests for it.)
+
+### Kotlin code
+
+The Kotlin wrapper code for a component should have its own test suite, which should follow the general guidelines for
+[testing Android code in Mozilla projects](https://github.com/mozilla-mobile/shared-docs/blob/master/android/testing.md#jvm-testing).
+In practice that means we use
+[JUnit](https://github.com/mozilla-mobile/shared-docs/blob/master/android/testing.md#junit-testing-framework)
+as the test framework and
+[Robolectric](https://github.com/mozilla-mobile/shared-docs/blob/master/android/testing.md#robolectric-android-api-shadows)
+to provide implementations of Android-specific APIs.
+
+The Kotlin tests for a component should be runnable via `./gradlew <component>:test`.
+
+The tests at this layer are designed to ensure that the API binding code is working as intended,
+and should not repeat tests for functionality that is already well tested at the Rust level.
+But given that the Kotlin bindings involve a non-trivial amount of hand-written boilerplate code,
+it's important to exercise that code throughly.
+
+One complication with running Kotlin tests is that the code needs to run on your local development machine,
+but the Kotlin code's native dependencies are typically compiled and packaged for Android devices. The
+tests need to ensure that an appropriate version of JNA and of the compiled Rust code is available in
+their library search path at runtime. Our `build.gradle` files contain a collection of hackery that ensures
+this, which should be copied into any new components.
+
+(Editor's note: I remain hopeful that one day we'll autogenerate most of the Kotlin binding code, and in
+such a world we don't need to invest in tests for it.)
+
+XXX TODO: talk about proguard? I don't really understand it...
+
+XXX TODO: any additional tips here, such as mocking out storage etc?
```

Comment:
I think we've decided on using in-memory DBs, by specifying the path as either `:memory:` or `file:some-name?mode=memory`, for other components—the logins component is an exception; it has a `MemoryLoginsStorage` class instead. Not sure if that would be a useful tip or if it belongs here, though.

Reply:
We've deprecated `MemoryLoginsStorage` FWIW